admin/code/CampaignAdmin.php

```diff
@@ -0,0 +1,408 @@
+			"attributes": [],
+			"data": []
+		}, {
+			"name": "ChangeSets",
+			"id": "Form_EditForm_ChangeSets",
+			"type": "Custom",
+			"component": "GridField",
+			"holder_id": null,
+			"title": "Campaigns",
+			"source": null,
+			"extraClass": null,
+			"description": null,
+			"rightTitle": null,
+			"leftTitle": null,
+			"readOnly": false,
+			"disabled": false,
+			"customValidationMessage": "",
+			"attributes": [],
+			"data": {
+				"recordType": "ChangeSet",
+				"collectionReadEndpoint": {
+					"url": "admin\/campaigns\/sets",
+					"method": "GET"
+				},
+				"itemReadEndpoint": {
+					"url": "admin\/campaigns\/set\/:id",
+					"method": "GET"
+				},
+				"itemUpdateEndpoint": {
+					"url": "admin\/campaigns\/set\/:id",
+					"method": "PUT"
+				},
+				"itemCreateEndpoint": {
+					"url": "admin\/campaigns\/set\/:id",
+					"method": "POST"
+				},
+				"itemDeleteEndpoint": {
+					"url": "admin\/campaigns\/set\/:id",
+					"method": "DELETE"
+				},
+				"editFormSchemaEndpoint": "admin\/campaigns\/schema\/DetailEditForm",
+				"columns": [
+					{"name": "Title", "field": "Name"},
+					{"name": "Changes", "field": "_embedded.ChangeSetItems.length"},
+					{"name": "Description", "field": "Description"}
+				]
+			}
+		}, {
+			"name": "SecurityID",
+			"id": "Form_EditForm_SecurityID",
+			"type": "Hidden",
+			"component": null,
+			"holder_id": null,
+			"title": "Security ID",
+			"source": null,
+			"extraClass": "hidden",
+			"description": null,
+			"rightTitle": null,
+			"leftTitle": null,
+			"readOnly": false,
+			"disabled": false,
+			"customValidationMessage": "",
+			"attributes": [],
+			"data": []
+		}],
+		"actions": []
+	}
+}
+JSON;
+
+		$formName = $request->param('ID');
+		if($formName == 'EditForm') {
+			$response = $this->getResponse();
+			$response->addHeader('Content-Type', 'application/json');
+			$response->setBody($json);
+			return $response;
+		} else {
+			return parent::schema($request);
+		}
+	}
+
+	/**
+	 * REST endpoint to create a campaign.
+	 *
+	 * @param SS_HTTPRequest $request
+	 *
+	 * @return SS_HTTPResponse
+	 */
+	public function createCampaign(SS_HTTPRequest $request) {
+		$response = new SS_HTTPResponse();
+		$response->addHeader('Content-Type', 'application/json');
+		$response->setBody(Convert::raw2json(['campaign' => 'create']));
+
+		return $response;
+	}
+
+	/**
+	 * REST endpoint to get a list of campaigns.
+	 *
+	 * @param SS_HTTPRequest $request
+	 *
+	 * @return SS_HTTPResponse
+	 */
+	public function readCampaigns(SS_HTTPRequest $request) {
+		$response = new SS_HTTPResponse();
+		$response->addHeader('Content-Type', 'application/json');
+		$hal = $this->getListResource();
+		$response->setBody(Convert::array2json($hal));
+		return $response;
+		}
+
+	/**
+	 * Get list contained as a hal wrapper
+	 *
+	 * @return array
+	 */
+	protected function getListResource() {
+		$items = $this->getListItems();
+		$count = $items->count();
+		$hal = [
```

Comment:
Similar to HTTPResponse, this is a common repeated pattern looking to be abstracted.